pkg/archive/changes.go

```diff
@@ -123,8 +127,16 @@ func Changes(layers []string, rw string) ([]Change, error) {
 							return nil
 						}
 					}
+					changedDirs[path] = struct{}{}
 					change.Kind = ChangeModify
 					break
+				} else if os.IsNotExist(err) && f.Mode().IsRegular() {
```

Comment:
Also, if I'm following all of this correctly, we're trying to fix the case where we walk to a file but not its parent dir (weird, but ok), right? If so, it seems like:
1 - we should check to see if all parents are in the change list, not just the first parent dir
2 - we should check irrespective of whether 'f' is a file or a dir - if there's a really a bug that makes it we can walk to a file w/o seeing its dir, then it seems like we could also walk to a dir w/o see its parent dir too.
I guess what I'm saying is.... should this code here be rewritten as:
```
tmpChanges := []Change{}
for parent := filepath.Dir( path ) ; parent != "/" ; parent = filepath.Dir(parent) {
  if changedDirs[parent] != nil {
    break
  }
  tmpChanges = append(tmpChanges, Change{Path: parent, Kind: ChangeModify}
  changedDirs[parent] = struct{}{}
}
for i := len(tmpChanges) ; i > 0 ; i-- {
  changes = append(changed, tmpChanges[i-1]
}
changedDirs[path] = struct{}{}
```
and moved to right before the `changes = append(changes, change)` line ?

But I may not be following what going on  :-)